File: backend/pkg/database/repository/tag.go
package repository

import (
	"context"
	"strings"

	"backend/pkg/core"
	"backend/pkg/database/models"
	"backend/pkg/types"

	"gorm.io/gorm"
)

// TagRepository provides data access for tags
type TagRepository struct {
	db *gorm.DB
}

// NewTagRepository creates a new TagRepository
func NewTagRepository(db *gorm.DB) *TagRepository {
	return &TagRepository{db: db}
}

// tagModelToType converts a database model to types
func tagModelToType(m *models.Tag) *types.Tag {
	return &types.Tag{
		ID:        m.ID,
		Name:      m.Name,
		CreatedAt: m.CreatedAt,
	}
}

// tagTypeToModel converts a types type to database model
func tagTypeToModel(t *types.Tag) *models.Tag {
	return &models.Tag{
		ID:        t.ID,
		Name:      t.Name,
		CreatedAt: t.CreatedAt,
	}
}

// FindByID retrieves a tag by its ID
func (r *TagRepository) FindByID(ctx context.Context, id int) (*types.Tag, error) {
	var model models.Tag
	if err := r.db.WithContext(ctx).First(&model, id).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return tagModelToType(&model), nil
}

// FindByName retrieves a tag by its name (case-insensitive)
func (r *TagRepository) FindByName(ctx context.Context, name string) (*types.Tag, error) {
	var model models.Tag
	if err := r.db.WithContext(ctx).Where("LOWER(name) = LOWER(?)", name).First(&model).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, core.ErrNotFound
		}
		return nil, err
	}
	return tagModelToType(&model), nil
}

// FindByIDs retrieves tags by their IDs
func (r *TagRepository) FindByIDs(ctx context.Context, ids []int64) ([]types.Tag, error) {
	var tagModels []models.Tag
	if err := r.db.WithContext(ctx).Where("id IN ?", ids).Find(&tagModels).Error; err != nil {
		return nil, err
	}

	tags := make([]types.Tag, len(tagModels))
	for i, m := range tagModels {
		tags[i] = *tagModelToType(&m)
	}
	return tags, nil
}

// EnsureExists creates tags if they don't exist and returns their IDs
func (r *TagRepository) EnsureExists(ctx context.Context, names []string) ([]int64, error) {
	tagIDs := make([]int64, 0, len(names))

	for _, name := range names {
		name = strings.TrimSpace(name)
		if name == "" {
			continue
		}

		var existingTag models.Tag
		err := r.db.WithContext(ctx).Where("LOWER(name) = LOWER(?)", name).First(&existingTag).Error
		if err == gorm.ErrRecordNotFound {
			// Create new tag
			newTag := &models.Tag{Name: name}
			if err := r.db.WithContext(ctx).Create(newTag).Error; err != nil {
				return nil, err
			}
			tagIDs = append(tagIDs, int64(newTag.ID))
		} else if err != nil {
			return nil, err
		} else {
			tagIDs = append(tagIDs, int64(existingTag.ID))
		}
	}

	return tagIDs, nil
}

// List retrieves all tags
func (r *TagRepository) List(ctx context.Context) ([]types.Tag, error) {
	var tagModels []models.Tag
	if err := r.db.WithContext(ctx).Order("name ASC").Find(&tagModels).Error; err != nil {
		return nil, err
	}

	tags := make([]types.Tag, len(tagModels))
	for i, m := range tagModels {
		tags[i] = *tagModelToType(&m)
	}
	return tags, nil
}

// Save creates or updates a tag
func (r *TagRepository) Save(ctx context.Context, tag *types.Tag) error {
	model := tagTypeToModel(tag)
	if model.ID == 0 {
		// Create new tag
		if err := r.db.WithContext(ctx).Create(model).Error; err != nil {
			return err
		}
		tag.ID = model.ID
		return nil
	}

	// Update existing tag
	return r.db.WithContext(ctx).Save(model).Error
}

// Delete removes a tag by its ID
func (r *TagRepository) Delete(ctx context.Context, id int) error {
	result := r.db.WithContext(ctx).Delete(&models.Tag{}, id)
	if result.Error != nil {
		return result.Error
	}
	if result.RowsAffected == 0 {
		return core.ErrNotFound
	}
	return nil
}

File: backend/src/database/repository/tag.rs
use async_trait::async_trait;
use diesel::{
    ExpressionMethods, OptionalExtension, PgArrayExpressionMethods, QueryDsl, SelectableHelper,
    dsl::count_star,
    result::{DatabaseErrorKind, Error as DieselError},
    sql_types::Text,
};
use diesel_async::{AsyncConnection, RunQueryDsl};

use crate::{
    core::tag::{Tag, TagRepository},
    database::{
        models::tag::{NewTagRow, TagRow},
        pool::PgPool,
    },
    error::AppError,
    schema::{article, project, tag},
};

diesel::define_sql_function!(fn lower(value: Text) -> Text);

#[derive(Clone)]
pub struct DieselTagRepository {
    pool: PgPool,
}

impl DieselTagRepository {
    pub const fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    async fn connection(
        &self,
    ) -> Result<
        diesel_async::pooled_connection::deadpool::Object<diesel_async::AsyncPgConnection>,
        AppError,
    > {
        self.pool.get().await.map_err(|_| AppError::Database)
    }
}

#[async_trait]
impl TagRepository for DieselTagRepository {
    async fn find_by_id(&self, id: i32) -> Result<Tag, AppError> {
        let mut connection = self.connection().await?;
        tag::table
            .find(id)
            .select(TagRow::as_select())
            .first::<TagRow>(&mut connection)
            .await
            .optional()
            .map_err(map_diesel_error)?
            .map(Tag::from)
            .ok_or(AppError::NotFound)
    }

    async fn find_by_name(&self, name: &str) -> Result<Tag, AppError> {
        let mut connection = self.connection().await?;
        tag::table
            .filter(lower(tag::name).eq(lower(name)))
            .order(tag::id.asc())
            .select(TagRow::as_select())
            .first::<TagRow>(&mut connection)
            .await
            .optional()
            .map_err(map_diesel_error)?
            .map(Tag::from)
            .ok_or(AppError::NotFound)
    }

    async fn find_by_ids(&self, ids: &[i64]) -> Result<Vec<Tag>, AppError> {
        let ids = ids
            .iter()
            .copied()
            .map(|id| {
                i32::try_from(id)
                    .map_err(|_| AppError::InvalidInput("tag ID is out of range".to_owned()))
            })
            .collect::<Result<Vec<_>, _>>()?;
        let mut connection = self.connection().await?;
        tag::table
            .filter(tag::id.eq_any(ids))
            .order(tag::id.asc())
            .select(TagRow::as_select())
            .load::<TagRow>(&mut connection)
            .await
            .map(|rows| rows.into_iter().map(Tag::from).collect())
            .map_err(map_diesel_error)
    }

    async fn ensure_exists(&self, names: &[String]) -> Result<Vec<i64>, AppError> {
        let mut connection = self.connection().await?;
        connection
            .transaction::<Vec<i64>, DieselError, _>(async |connection| {
                let mut ids = Vec::with_capacity(names.len());
                for raw_name in names {
                    let name = raw_name.trim();
                    if name.is_empty() {
                        continue;
                    }

                    diesel::sql_query(
                        "SELECT pg_advisory_xact_lock(hashtextextended(LOWER($1), 0))",
                    )
                    .bind::<Text, _>(name)
                    .execute(connection)
                    .await?;

                    let existing = tag::table
                        .filter(lower(tag::name).eq(lower(name)))
                        .order(tag::id.asc())
                        .select(TagRow::as_select())
                        .first::<TagRow>(connection)
                        .await
                        .optional()?;

                    let id = if let Some(existing) = existing {
                        existing.id
                    } else {
                        diesel::insert_into(tag::table)
                            .values(NewTagRow { name })
                            .returning(tag::id)
                            .get_result::<i32>(connection)
                            .await?
                    };
                    ids.push(i64::from(id));
                }
                Ok(ids)
            })
            .await
            .map_err(map_diesel_error)
    }

    async fn list(&self) -> Result<Vec<Tag>, AppError> {
        let mut connection = self.connection().await?;
        tag::table
            .order(tag::name.asc())
            .select(TagRow::as_select())
            .load::<TagRow>(&mut connection)
            .await
            .map(|rows| rows.into_iter().map(Tag::from).collect())
            .map_err(map_diesel_error)
    }

    async fn save(&self, value: &mut Tag) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        if value.id == 0 {
            let row = diesel::insert_into(tag::table)
                .values(NewTagRow { name: &value.name })
                .returning(TagRow::as_returning())
                .get_result::<TagRow>(&mut connection)
                .await
                .map_err(map_diesel_error)?;
            value.id = row.id;
            return Ok(());
        }

        let affected = diesel::update(tag::table.find(value.id))
            .set(tag::name.eq(&value.name))
            .execute(&mut connection)
            .await
            .map_err(map_diesel_error)?;
        if affected == 0 {
            Err(AppError::NotFound)
        } else {
            Ok(())
        }
    }

    async fn delete(&self, id: i32) -> Result<(), AppError> {
        let mut connection = self.connection().await?;
        let affected = diesel::delete(tag::table.find(id))
            .execute(&mut connection)
            .await
            .map_err(map_diesel_error)?;
        if affected == 0 {
            Err(AppError::NotFound)
        } else {
            Ok(())
        }
    }

    async fn is_used(&self, id: i32) -> Result<bool, AppError> {
        let mut connection = self.connection().await?;
        let article_count = article::table
            .filter(article::tag_ids.contains(vec![Some(id)]))
            .select(count_star())
            .first::<i64>(&mut connection)
            .await
            .map_err(map_diesel_error)?;
        if article_count > 0 {
            return Ok(true);
        }
        project::table
            .filter(project::tag_ids.contains(vec![Some(id)]))
            .select(count_star())
            .first::<i64>(&mut connection)
            .await
            .map(|count| count > 0)
            .map_err(map_diesel_error)
    }
}

fn map_diesel_error(error: DieselError) -> AppError {
    match error {
        DieselError::NotFound => AppError::NotFound,
        DieselError::DatabaseError(DatabaseErrorKind::UniqueViolation, _) => {
            AppError::Conflict("resource already exists".to_owned())
        }
        _ => AppError::Database,
    }
}
